File: android/src/main/java/com/lebrislo/bluetooth/mesh/plugin/ConfigPluginCall.kt
package com.lebrislo.bluetooth.mesh.plugin

import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.PluginCall
import no.nordicsemi.android.mesh.transport.ConfigAppKeyStatus
import no.nordicsemi.android.mesh.transport.ConfigCompositionDataStatus
import no.nordicsemi.android.mesh.transport.ConfigModelAppStatus
import no.nordicsemi.android.mesh.transport.ConfigNodeResetStatus
import no.nordicsemi.android.mesh.transport.MeshMessage

/**
 * This class is used to generate a response for a Config plugin call.
 */
class ConfigPluginCall(val meshOperationCallback: Int, val meshAddress: Int, call: PluginCall) : BasePluginCall(call) {
    companion object {
        /**
         * Generates a response for a Config plugin call.
         *
         * @param meshMessage Mesh message.
         */
        @JvmStatic
        fun generateConfigPluginCallResponse(meshMessage: MeshMessage): JSObject {
            val result = JSObject()
            result.put("src", meshMessage.src)
            result.put("dst", meshMessage.dst)
            result.put("opcode", meshMessage.opCode)
            result.put(
                "data", when (meshMessage) {
                    is ConfigAppKeyStatus -> configAppKeyStatusResponse(meshMessage)
                    is ConfigNodeResetStatus -> configNodeResetStatusResponse(meshMessage)
                    is ConfigModelAppStatus -> configModelAppStatusResponse(meshMessage)
                    is ConfigCompositionDataStatus -> configCompositionDataStatusResponse(meshMessage)
                    else -> JSObject()
                }
            )
            return result
        }

        private fun configAppKeyStatusResponse(meshMessage: ConfigAppKeyStatus): JSObject {
            val data = JSObject()
            data.put("status", meshMessage.statusCode)
            data.put("netKeyIndex", meshMessage.netKeyIndex)
            data.put("appKeyIndex", meshMessage.appKeyIndex)
            return data
        }

        private fun configNodeResetStatusResponse(meshMessage: ConfigNodeResetStatus): JSObject {
            val data = JSObject()
            data.put("status", meshMessage.statusCode)
            return data
        }

        private fun configModelAppStatusResponse(meshMessage: ConfigModelAppStatus): JSObject {
            val data = JSObject()
            data.put("status", meshMessage.statusCode)
            data.put("elementAddress", meshMessage.elementAddress)
            data.put("modelId", meshMessage.modelIdentifier)
            data.put("appKeyIndex", meshMessage.appKeyIndex)
            return data
        }

        private fun configCompositionDataStatusResponse(meshMessage: ConfigCompositionDataStatus): JSObject {
            val data = JSObject()
            data.put("status", meshMessage.statusCode)
            data.put("companyIdentifier", meshMessage.companyIdentifier)
            data.put("productIdentifier", meshMessage.productIdentifier)
            data.put("versionIdentifier", meshMessage.versionIdentifier)
            data.put("crpl", meshMessage.crpl)
            data.put("features", meshMessage.features)
            data.put("relayFeatureSupported", meshMessage.isRelayFeatureSupported)
            data.put("proxyFeatureSupported", meshMessage.isProxyFeatureSupported)
            data.put("friendFeatureSupported", meshMessage.isFriendFeatureSupported)
            data.put("lowPowerFeatureSupported", meshMessage.isLowPowerFeatureSupported)

            val elements = JSArray()
            meshMessage.elements.forEach { element ->
                val elementData = JSObject()
                elementData.put("locationDescriptor", element.value.locationDescriptor)
                elementData.put("elementAddress", element.value.elementAddress)

                val models = JSArray()
                element.value.meshModels.forEach { model ->
                    val modelData = JSObject()
                    modelData.put("modelId", model.value.modelId)
                    modelData.put("modelName", model.value.modelName)
                    models.put(modelData)
                }
                elementData.put("models", models)

                elements.put(elementData)
            }

            data.put("elements", elements)

            return data
        }
    }
}

File: ios/Sources/BluetoothMeshPlugin/plugin/ConfigPluginCall.swift
////
////  ConfigPluginCall.swift
////  BluetoothMeshPlugin
////
////  Created by LE BRIS Loris on 22/10/2025.
////
//
//import Foundation
//import Capacitor
//
//class ConfigPluginCall: BasePluginCall {
//    let meshOperationCallback: Int
//    let meshAddress: Int
//
//    init(meshOperationCallback: Int, meshAddress: Int, call: CAPPluginCall) {
//        self.meshOperationCallback = meshOperationCallback
//        self.meshAddress = meshAddress
//        super.init(call: call)
//    }
//
//    static func generateConfigPluginCallResponse(meshMessage: MeshMessage) -> PluginCallResultData {
//        var result = PluginCallResultData()
//        result["src"] = meshMessage.
//        result["dst"] = meshMessage.dst
//        result["opcode"] = meshMessage.opCode
//
//        if let msg = meshMessage as? ConfigAppKeyStatus {
//            result["data"] = configAppKeyStatusResponse(msg)
//        } else if let msg = meshMessage as? ConfigNodeResetStatus {
//            result["data"] = configNodeResetStatusResponse(msg)
//        } else if let msg = meshMessage as? ConfigModelAppStatus {
//            result["data"] = configModelAppStatusResponse(msg)
//        } else if let msg = meshMessage as? ConfigCompositionDataStatus {
//            result["data"] = configCompositionDataStatusResponse(msg)
//        } else {
//            result["data"] = PluginCallResultData()
//        }
//        return result
//    }
//
//    private static func configAppKeyStatusResponse(_ meshMessage: ConfigAppKeyStatus) -> PluginCallResultData {
//        var data = PluginCallResultData()
//        data["status"] = meshMessage.status
//        data["netKeyIndex"] = meshMessage.networkKeyIndex
//        data["appKeyIndex"] = meshMessage.applicationKeyIndex
//        return data
//    }
//
//    private static func configNodeResetStatusResponse(_ meshMessage: ConfigNodeResetStatus) -> PluginCallResultData {
//        var data = PluginCallResultData()
//        data["status"] = meshMessage.status
//        return data
//    }
//
//    private static func configModelAppStatusResponse(_ meshMessage: ConfigModelAppStatus) -> PluginCallResultData {
//        var data = PluginCallResultData()
//        data["status"] = meshMessage.status
//        data["elementAddress"] = meshMessage.elementAddress
//        data["modelId"] = meshMessage.modelIdentifier
//        data["appKeyIndex"] = meshMessage.applicationKeyIndex
//        return data
//    }
//
//    private static func configCompositionDataStatusResponse(_ meshMessage: ConfigCompositionDataStatus) -> PluginCallResultData {
//        var data = PluginCallResultData()
//        data["status"] = meshMessage
//        data["companyIdentifier"] = meshMessage.companyIdentifier
//        data["productIdentifier"] = meshMessage.productIdentifier
//        data["versionIdentifier"] = meshMessage.versionIdentifier
//        data["crpl"] = meshMessage.crpl
//        data["features"] = meshMessage.features
//        data["relayFeatureSupported"] = meshMessage.isRelayFeatureSupported
//        data["proxyFeatureSupported"] = meshMessage.isProxyFeatureSupported
//        data["friendFeatureSupported"] = meshMessage.isFriendFeatureSupported
//        data["lowPowerFeatureSupported"] = meshMessage.isLowPowerFeatureSupported
//
//        var elements: JSArray = []
//        for element in meshMessage.elements {
//            var elementData = PluginCallResultData()
//            elementData["locationDescriptor"] = element.value.locationDescriptor
//            elementData["elementAddress"] = element.value.elementAddress
//
//            var models: JSArray = []
//            for model in element.value.meshModels {
//                var modelData: PluginCallResultData = [:]
//                modelData["modelId"] = model.value.modelId
//                modelData["modelName"] = model.value.modelName
//                models.append(modelData)
//            }
//            elementData["models"] = models
//            elements.append(elementData)
//        }
//        data["elements"] = elements
//        return data
//    }
//}
